File: src/collectors/apps.ts
import type { Collector, CollectorResult } from "./types";
import { makeSection } from "./types";
import { type CommandEnv, defaultEnv } from "./env";

const RAYCAST_PLIST = "/Applications/Raycast.app/Contents/Info.plist";
const ALTTAB_PLIST = "/Applications/AltTab.app/Contents/Info.plist";

/** Parse `ls /Applications` output (one entry per line). */
export function parseAppList(text: string): string[] {
  return text
    .trim()
    .split("\n")
    .map((a) => a.trim())
    .filter(Boolean)
    .sort();
}

export function makeAppsCollector(env: CommandEnv = defaultEnv): Collector {
  return async () => {
    if (process.platform !== "darwin") return {};
    const result: CollectorResult = {};

    result["apps.raycast"] = makeSection("apps.raycast", {
      pairs: { installed: (await env.fileExists(RAYCAST_PLIST)) ? "true" : "false" },
    });

    const alttabInstalled = await env.fileExists(ALTTAB_PLIST);
    const alttabPairs: Record<string, string> = {
      installed: alttabInstalled ? "true" : "false",
    };
    if (alttabInstalled) {
      try {
        const prefs = await env.run(["defaults", "read", "com.lwouis.alt-tab-macos"]);
        if (prefs.trim()) alttabPairs.preferences = "exists";
      } catch {}
    }
    result["apps.alttab"] = makeSection("apps.alttab", { pairs: alttabPairs });

    try {
      const apps = parseAppList(await env.run(["ls", "/Applications"]));
      if (apps.length) {
        result["apps.macos"] = makeSection("apps.macos", {
          items: apps.map((a) => ({ raw: a, columns: [a] })),
        });
      }
    } catch {}

    return result;
  };
}

export const collectApps = makeAppsCollector();

File: internal/collect/apps.go
package collect

import (
	"sort"
	"strings"

	"github.com/doguyilmaz/dothaven/internal/snapshot"
)

const (
	raycastPlist = "/Applications/Raycast.app/Contents/Info.plist"
	altTabPlist  = "/Applications/AltTab.app/Contents/Info.plist"
)

// ParseAppList parses `ls /Applications` output (one entry per line): trims each
// line, drops blanks, and returns the sorted set.
func ParseAppList(text string) []string {
	lines := strings.Split(strings.TrimSpace(text), "\n")
	out := make([]string, 0, len(lines))
	for _, l := range lines {
		if t := strings.TrimSpace(l); t != "" {
			out = append(out, t)
		}
	}
	sort.Strings(out)
	return out
}

// AppsCollector gathers the macOS application inventory.
func AppsCollector(c Ctx) snapshot.Snapshot {
	out := snapshot.Snapshot{}

	out["apps.raycast"] = snapshot.Section{
		Pairs: map[string]string{"installed": appsInstalled(c.Env.Exists(raycastPlist))},
	}

	altInstalled := c.Env.Exists(altTabPlist)
	altPairs := map[string]string{"installed": appsInstalled(altInstalled)}
	if altInstalled {
		prefs, _ := c.Env.Run(c.Context, "defaults", "read", "com.lwouis.alt-tab-macos")
		if strings.TrimSpace(prefs) != "" {
			altPairs["preferences"] = "exists"
		}
	}
	out["apps.alttab"] = snapshot.Section{Pairs: altPairs}

	listing, _ := c.Env.Run(c.Context, "ls", "/Applications")
	if apps := ParseAppList(listing); len(apps) > 0 {
		out["apps.macos"] = snapshot.Section{Items: toItems(apps)}
	}

	return out
}

func appsInstalled(b bool) string {
	if b {
		return "true"
	}
	return "false"
}
